invoices/update.go

```diff
@@ -99,8 +129,122 @@ func updateInvoice(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
 		}
 	}
 
-	// If the invoice is already canceled, there is no further checking to
-	// do.
+	if ctx.mpp == nil {
+		return updateLegacy(ctx, inv)
+	}
+
+	return updateMpp(ctx, inv)
+}
+
+// updateMpp is a callback for DB.UpdateInvoice that contains the invoice
+// settlement logic for mpp payments.
+func updateMpp(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
+	*channeldb.InvoiceUpdateDesc, updateResult, error) {
+
+	// Start building the accept descriptor.
+	acceptDesc := &channeldb.HtlcAcceptDesc{
+		Amt:          ctx.amtPaid,
+		Expiry:       ctx.expiry,
+		AcceptHeight: ctx.currentHeight,
+		MppTotalAmt:  ctx.mpp.TotalMsat(),
+	}
+
+	// Only accept payments to open invoices. This behaviour differs from
+	// non-mpp payments that are accepted even after the invoice is settled.
+	// Because non-mpp payments don't have a payment address, this is needed
+	// to thwart probing.
+	if inv.State != channeldb.ContractOpen {
+		return nil, resultInvoiceNotOpen, nil
+	}
+
+	// Check the payment address that authorizes the payment.
+	if ctx.mpp.PaymentAddr() != inv.Terms.PaymentAddr {
+		return nil, resultAddressMismatch, nil
+	}
+
+	// Don't accept zero-valued sets.
+	if ctx.mpp.TotalMsat() == 0 {
+		return nil, resultHtlcSetTotalTooLow, nil
+	}
+
+	// Check that the total amt of the htlc set is high enough. In case this
+	// is a zero-valued invoice, it will always be enough.
+	if ctx.mpp.TotalMsat() < inv.Terms.Value {
+		return nil, resultHtlcSetTotalTooLow, nil
+	}
+
+	// Check whether total amt matches other htlcs in the set.
+	var newSetTotal lnwire.MilliSatoshi
+	for _, htlc := range inv.Htlcs {
+		// Only consider accepted mpp htlcs.
+		if htlc.State != channeldb.HtlcStateAccepted {
```

Comment:
could there ever be any other htlc states in the set at this point?

Reply:
Yes, htlcs that were canceled previously because they timed out.

Reply:
Comment added